wordpress-mobile/WordPress-iOS
# Feature/sign in refactor

Closes #4912

This PR refactors the existing `LoginViewController` and implements a new "magic-links" sign-in feature.  The new controllers are written in Swift, and their UI is mainly defined in storyboard. View controllers are responsible for a single form, reducing the amount of state that needs to be maintained and tested.  Common functionality has been factored out into helper classes and protocols as much as possible, leaving controllers with a common, informal interface but no (or very limited) duplication. 


To test:
The `LoginViewController` supported a variety of sign-in scenarios, so there is a lot to test.  For each of the following scenarios, test the primary use-case, optional features on the form (various buttons) and error states (offline, invalid form entry). 

Scenario 1: 
Sign in to wpcom with an email address using the magic link flow.  To test the final step, reopening the app passing the auth token, you'll need to copy the `wordpress://` link from safari and change it to `wpdebug://`. 

Scenario 2: 
Sign in to wpcom with a username and password. No 2fa. 
 
Scenario 3: 
Sign in to wpcom with a username and password with 2fa.

Scenario 4: 
Sign in via a self-hosted blog. 

Scenario 5:
Confirm 1Password integration by testing on a device with 1Password installed.  Confirm that the 1Password button is present on the email, username/password, and self-hosted forms and that it functions on each. 
Note that on the self-hosted form you should be prompted to enter an URL if that field is empty. 

Scenario 6:
Confirm Safari stored password integration. 
Store credentials to wpcom 
- Settings > Safari > AutoFill > Names and Passwords.  
- Settings > Safari > Passwords > Add Password for wordpress.com
Launch and confirm you're prompted, and asked to use stored credentials. 
Choose credentials and confirm that they are filled out and you end up on the username/password form. 
Try again, and dismiss the prompt by clicking "Not Now".  Then re-open the prompt by clicking the saved passwords button on the Email form (first screen).


Needs review: @koke
